use `format.any` to provide respond_to defaults

wjordan opened the pull request:
This PR should fix the `ActionController::UnknownFormat` errors when the `accept` header is not provided by browser/proxy (e.g. when it's filtered by CloudFront).

@laurelfan can you think of any gotchas for making this sort of change or should it be generally safe? I don't think we're expecting explicit `accept` values in request-header behavior for these cases, are we?

Also, I'm not sure if the existing unit test suite covers the case of a missing `accept` header- would we need to add an integration test for this?

laurelfan commented on dashboard/app/controllers/gallery_activities_controller.rb:
```diff
@@ -42,10 +42,10 @@ def create
 
       respond_to do |format|
         if @gallery_activity.save
-          format.json { render action: 'show', status: :created, location: @gallery_activity }
```
this one is fine, we only expect to call it with json

laurelfan commented on dashboard/app/controllers/registrations_controller.rb:
```diff
@@ -32,10 +32,10 @@ def update
           end
         end
-        format.json { head :no_content }
+        format.any { head :no_content }
```
this will break the edit account page because it's expecting two different types of requests -- json and regular html post -- and wants to handle them differently.

(the two different callers are the edit account page and the age interstitial)

Why can't we just pass the accept header through? These are also all posts so are uncacheable.

laurelfan replied:
Repro steps:

Age interstitial:
1. Log in as a teacher
2. Go to http://localhost.code.org:3000/teacher-dashboard/#/sections
- create a section with login type picture
- create a student without specifying age
- click show secret to see their secret picture
  1. Open the /section url for that section in a new incognito window
- log in as a student
- see age interstitial
- select age and click continue
  Expected: interstitial goes away
  Current behavior on prod: nothing appears to happen when clicking Continue

Edit account:
1. Sign in as any account
2. Go to "Edit Account" in user menu: http://localhost-studio.code.org:3000/users/edit
3. Change Display Name
4. Click [Update]
Expected (and current behavior on prod): Go back to edit account page with a success message at the top in green

Edit account with error:
1. Sign in as any account
2. Go to "Edit Account" in user menu: http://localhost-studio.code.org:3000/users/edit
3. Change Email to "a@a" (something that passes the js regex to enable the submit button but isn't actually a valid email)
4. Click [Update]
Expected (and current behavior on prod): Go back to edit account page with error messages in red and invalid fields highlighted

laurelfan commented:
Re: testing:

We should probably set the Accept header in all unit tests to whatever cloudfront sends (nothing?)

Here are all the tests that are currently setting the Accept header:

```
test/controllers/home_controller_test.rb:    @request.headers["Accept"] = "image/*"
test/controllers/home_controller_test.rb:    @request.headers["Accept"] = "image/*"
test/controllers/ops/cohorts_controller_test.rb:      @request.headers['Accept'] = 'application/json'
test/controllers/ops/districts_controller_test.rb:      @request.headers['Accept'] = 'text/html'
test/controllers/ops/segments_controller_test.rb:      @request.headers['Accept'] = 'application/json'
test/controllers/ops/workshop_attendance_controller_test.rb:      @request.headers['Accept'] = 'application/json'
test/controllers/ops/workshops_controller_test.rb:      @request.headers['Accept'] = 'application/json'
test/controllers/user_levels_controller_test.rb:    @request.headers['Accept'] = 'application/json'
test/controllers/user_levels_controller_test.rb:    @request.headers['Accept'] = 'application/json'
test/controllers/user_levels_controller_test.rb:    @request.headers['Accept'] = 'application/json'
```

Do these tests still work when we remove these lines? If not there might be more things that need to be fixed.

Also, the "update over 13 student with plaintext email" test in RegistrationsControllerTest is testing the format.html version of the RegistrationsController#update that you modified, does that pass if you remove the accept header (which is html by default?)

The "update student with age" test in RegistrationsControllerTest is intended to test the age interstitial code path -- however it's missing an assertion of what the response is. You can probably just add `assert_response :no_content` to that

wjordan commented:
thanks laurel for the detailed suggestions, i'll work on these tests.

We could pass the `accept` header through CloudFront as an alternative if adjusting the Rails code to not depend on the `accept` header value isn't feasible. My strong preference is to tweak the application rather than the http cache configuration though if at all possible though, since the latter has these drawbacks:
- more difficult to port across proxy-cache implementations
- slower and riskier to deploy
- more complex and error-prone 
- harder to test/debug across environments due to divergent implementations

laurelfan commented:
That makes sense -- I think if we can make the unit test environment (and maybe even the developer server setup) better match cloudfront by removing the accept headers that will really help testability.

In this case my guess is that the only application change that is required is doing something besides format/Accept header to distinguish the two different usages of RegistrationController#update (your tests will confirm/deny that). I think we can do that by using a file extension (like .json) to specify the format (probably worth manually testing this on staging and not just relying on unit tests).

In places where we don't actually use multiple formats in the same action (like the GalleryActivity stuff) I think we can just remove the format code (if the only format is format.any that doesn't really do anything).

I don't know whether the LevelSourceHintsController stuff is actually called with different formats; that looks like the rails generated scaffold so it's possible that it's not.

wjordan commented:
On the RegistrationsController, I think the current change should work fine because of the way Rails's accept-inference is implemented - in the absence of an `Accept` header it uses `js` as the default if the request is an XmlHttpRequest, and it assumes `html` as the default otherwise. I think the reason is wasn't working before is because `js` is distinct from `json` as a request format.

(Unless we're sending XHR requests with html accept headers- i don't think we're doing this, are we?)

I looked at and updated all of the tests explicitly setting accept headers:

> `test/controllers/home_controller_test.rb:    @request.headers["Accept"] = "image/*"
> test/controllers/home_controller_test.rb:    @request.headers["Accept"] = "image/*"`

These are tests for normalizing malformed `Accept` request headers, so stripping the headers should fix these cases as well. I'll leave these tests in place though.

> `test/controllers/ops/cohorts_controller_test.rb:      @request.headers['Accept'] = 'application/json'
> test/controllers/ops/segments_controller_test.rb:      @request.headers['Accept'] = 'application/json'
> test/controllers/ops/workshop_attendance_controller_test.rb:      @request.headers['Accept'] = 'application/json'
> test/controllers/ops/workshops_controller_test.rb:      @request.headers['Accept'] = 'application/json'`

We explicitly set `request.format = :json` in `OpsControllerBase` which should handle all of these cases, so I've removed Accept headers from the tests.

> `test/controllers/ops/districts_controller_test.rb:      @request.headers['Accept'] = 'text/html'`

This one is testing `"just return json whatever you ask for"` so we can leave this one in. I've also added `format: :html` to be extra sure we're just returning json no matter what.

> `test/controllers/user_levels_controller_test.rb:    @request.headers['Accept'] = 'application/json'
> test/controllers/user_levels_controller_test.rb:    @request.headers['Accept'] = 'application/json'
> test/controllers/user_levels_controller_test.rb:    @request.headers['Accept'] = 'application/json'`

We're always returning json for these, so i've removed the entire `respond_to` clause in the controller and the explicit `@request.headers` setter lines in the tests.

Pushed a commit with these changes to the PR, please take another look.

laurelfan commented:
Thanks for updating all the tests, LGTM
